docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-Vasyl-Teliman.md

```diff
@@ -0,0 +1,342 @@
+
+[#3723](https://github.com/KhronosGroup/SPIRV-Tools/issues/3723): Implement alias analysis.
+
+[#3696](https://github.com/KhronosGroup/SPIRV-Tools/issues/3696): Introduce context-dependent facts.
+
+[#3725](https://github.com/KhronosGroup/SPIRV-Tools/issues/3725): Determine if the function is pure.
+
+### Challenges
+Some of the implemented transformations were quite challenging. Most of the challenges came from various rules of the SPIR-V language that
+must be satisfied. Specific transformations and various challenging implementation details are mentioned below.
+
+#### [#3477](https://github.com/KhronosGroup/SPIRV-Tools/pull/3477): Permute instructions in the basic block
+There are simple algorithms that produce a random permutation of the sequence of elements. However, they can't be used to permute instructions in the basic block.
+This is because:
+1. Domination rules must be satisfied. That being said, every definition must dominate (read 'precede') all of its users. There are exceptions to this rule but
+we won't discuss them here.
+2. Swapping some instructions might change the semantics of the program. This is usually the case with instructions that cause side-effects: deal with memory, synchronization etc.
+
+Thus, the transformation simply swaps two consecutive instructions instead of permuting a range of them. This approach simplifies the implementation
+and makes the code more readable. Moreover, it is still able to produce a random permutation of all instructions since the transformation can be applied
+many times to the same basic block.
+
+#### [#3692](https://github.com/KhronosGroup/SPIRV-Tools/pull/3692): Propagate instructions from the basic block into its successors
+This transformation is part of the idea of being able to permute instructions in the program.
+
+This transformation has the same challenges as the previous one and more.
+1. Domination rules must be satisfied. We must make sure that all dependencies of the original instruction (before the instruction is propagated) dominate all the
+propagated copies of the original instruction. In the case of this transformation, `dominates` does not mean `syntactically precedes` if the original instruction is in the loop.
+2. We can't always guarantee that the propagated copy of the original instruction dominates all users of the original instruction. Thus, we do not preserve the id of
+the original instruction and instead decide if we can replace all usages of that id with propagated ids.
+3. Some propagated instructions might cause side-effects (e.g. write to memory etc). Thus, we might change the semantics of the program if we try to propagate them.
+
+There seems to be no single simple approach to overcome all these challenges. Thus, the transformation contains a sequence of conditions that must be satisfied to make
+sure that the program's semantics remain the same.
```

Comment:
Maybe add an example of at least one of the conditions, and what it tries to avoid.

Reply:
Fixed in a597e81.